charmcraft/charm_builder.py

```diff
@@ -233,31 +239,48 @@ def _install_dependencies(self, staging_venv_dir):
             _process_run(["python3", "-m", "venv", str(staging_venv_dir)])
         pip_cmd = str(_find_venv_bin(staging_venv_dir, "pip3"))
 
-        with instrum.Timer("Getting pip version"):
-            _process_run([pip_cmd, "--version"])
+        with instrum.Timer("Ensuring a recent enough pip..."):
+            if get_pip_version(pip_cmd) < (22, 0):
+                _process_run([pip_cmd, "install", "-U", "pip"])
 
         with instrum.Timer("Installing all dependencies"):
-            if self.binary_python_packages:
-                # install python packages, allowing binary packages
-                cmd = [pip_cmd, "install", "--upgrade"]  # base command
-                cmd.extend(self.binary_python_packages)  # the python packages to install
-                _process_run(cmd)
-
-            if self.python_packages:
-                # install python packages from source
-                cmd = [pip_cmd, "install", "--upgrade", "--no-binary", ":all:"]  # base command
-                cmd.extend(self.python_packages)  # the python packages to install
-                _process_run(cmd)
-
-            if self.requirement_paths or self.charmlib_deps:
-                cmd = [pip_cmd, "install", "--upgrade", "--no-binary", ":all:"]  # base command
-
+            try:
+                _process_run(
+                    get_pip_command(
+                        [pip_cmd, "install"],
+                        self.requirement_paths,
+                        source_deps=[*self.python_packages, *self.charmlib_deps],
+                        binary_deps=self.binary_python_packages,
+                    )
+                )
+            except RuntimeError:
+                # Fall back to old behaviour if pip install fails.
+                # This will go away with a future base.
+                print(
+                    "WARNING: Initial package installation failed. "
+                    "Falling back to older method, which may leave your charm "
+                    "in an un-runnable state."
+                )
```

Comment:
Will this only show in verbose mode?

Reply:
The charm builder's content ends up getting output directly to the console regardless, but with `stdout` and `stderr` collapsed (so there's no point in setting `file=sys.stderr`).